user - Warn when an invalid shell is specified on BusyBox systems

samdoran opened the pull request:
##### SUMMARY

Only for BusyBox since that was the context where this initally came up. It may make sense to add this warning more broadly.

`chsh` on Alpine warns but does not error when changing to an invalid shell. Other distros error with an invalid shell.

Depends on #86143.

Example output
```
> ansible-playbook playbook.yml

PLAY [all] ***************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************
ok: [alpine]

TASK [user] **************************************************************************************************
[WARNING]: '/usr/bin/nope' is not listed as a valid shell on the remote host.
ok: [alpine]

PLAY RECAP ***************************************************************************************************
alpine                     : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

samdoran committed:
user - Warn when an invalid shell is specified

Only for BusyBox since that was the context where this initally came up. It may
make sense to add this warning more broadly.

chsh on Alpine warns but does not error when changing to an invalid shell.
Other distros error with an invalid shell.

samdoran committed:
user - Warn when an invalid shell is specified

Only for BusyBox since that was the context where this initally came up. It may
make sense to add this warning more broadly.

chsh on Alpine warns but does not error when changing to an invalid shell.
Other distros error with an invalid shell.

Signed-off-by: Sam Doran <sdoran@redhat.com>
Co-authored-by: Abhijeet Kasurde <Akasurde@redhat.com>

samdoran committed:
Add changelog fragment

KalleDK commented:
@samdoran just wanna say thanks for remembering and implementing this  👍 🥇

mkrizek commented on lib/ansible/modules/user.py:
```diff
@@ -3114,7 +3114,12 @@ class BusyBox(User):
     """
+    def _validate_shell(self):
+        with open("/etc/shells", "rb") as f:
+            shells = [line.decode().strip() for line in f.readlines() if not line.startswith(b"#")]
```
We should probably filter blank lines too (the change assumes file open in text mode which is how I tested it):

```suggestion
            shells = [l for line in f.readlines() if (l := line.strip()) and not l.startswith("#")]
```

samdoran replied:
Sure thing!

samdoran replied:
I went with a nested comprehension using a generator just to keep the module more portable.

mkrizek commented on lib/ansible/modules/user.py:
```diff
@@ -3114,7 +3114,12 @@ class BusyBox(User):
         - modify_user()
     """
+    def _validate_shell(self):
+        with open("/etc/shells", "rb") as f:
```
I am curious, do we need to open in binary mode here?

samdoran replied:
I did that mainly to be consistent with other `open` operations in the file. I _think_ it would be ok to open in text mode but I confess I don't know for sure.

mkrizek commented on lib/ansible/modules/user.py:
```diff
@@ -3114,7 +3114,12 @@ class BusyBox(User):
+        with open("/etc/shells", "rb") as f:
+            shells = [line.decode().strip() for line in f.readlines() if not line.startswith(b"#")]
 
+        if self.shell not in shells:
```
Can we reach this point even if the shell is not provided by the user? If so we should not run this check, right?

samdoran replied:
Good point. Yes, this could be reached if the `shell` param isn't provided, in which case it would be an empty string. I need to look into that a bit more.

samdoran replied:
I'll return early if `self.shell` is empty.

samdoran replied:
Looks like the current tests caught this.

samdoran committed:
Return early if shell was not provided as a parameter

Strip blank lines in addition to comments when looking at /etc/shells.

Use a nested comprehension instead of the assignment operator so that the
module will work with older Python versions.

Open the file in text mode since I can’t think of a reason it would be unsafe
to do so in this situation.

samdoran committed:
Also validate shell when creating the user account

mkrizek commented on lib/ansible/modules/user.py:
```diff
@@ -3132,6 +3132,21 @@ class BusyBox(User):
+        if not self.shell:
+            return
+
+        with open("/etc/shells", "r") as f:
```
I understand `/etc/shells` is not POSIX, should we catch `FileNotFoundError`, silently ignore it and just `return` in such a case?

But it is just me being paranoid, I suspect it is unlikely the file won't be present.

samdoran replied:
That's probably not a bad idea. Defensive programming is a good thing.

mkrizek commented on lib/ansible/modules/user.py:
```diff
@@ -3132,6 +3132,21 @@ class BusyBox(User):
+        with open("/etc/shells", "r") as f:
+            shells = [
+                shell
+                for shell in (line.strip() for line in f.readlines())
```
```suggestion
                for shell in (line.strip() for line in f)
```

It doesn't make a difference in this scenario, just sharing something I  learned only recently that iterating over the file object is more performant as `readlines()` loads the whole file into memory.

samdoran replied:
That actually makes way more sense here. I'm so used to opening a file and storing the contents that I forgot the `readlines()` call isn't strictly necessary for this use. And it's much more efficient as you pointed out. Thanks!

mkrizek reviewed (approved):
I had two comments that don't need to be addressed. LGTM

samdoran committed:
Handle when /etc/shells does not exist

samdoran committed:
Iterate over the file object

Since the file is being iterated over already, calling readlines() is
unnecessary. It also increases memory usage by reading the file into memory
and then iterating over that.
